### DISCUSSION TITLE
Syntax highlighting and navigation on Java code

### DISCUSSION THREAD

**Discussion by yadavan88** [VirtusLab/scala-cli#2995] [Education]
Hi,
I know that Scala CLI is definitely not intended for Java projects. 
However, recently, I am trying to encourage Java devs to Scala, especially by introducing the powers of Scala 3 along with Scala-CLI and I got really nice feedback. 
In one case, I went one step ahead and used Java code in my discussions/meetups with friends and colleagues. And some people were really interested and curious about the usage of Scala-CLI for writing Java code initially and slowly learning Scala along with it. I just published a new [blog](https://yadukrishnan.live/developing-java-applications-with-scala-cli) about it as well
However, I was wondering if there is any way to get the navigation and code highlighting with third party deps. With core-java classes, it works fine with java plugins in intellij and vscode. 
IS there any quick way or hack that makes auto-complete and code navigation for Java code in Scala-CLI?

Thanks for working on this amazing tool!

**Comment by tgodzik** [comment]
Thanks for the feedback!

> However, I was wondering if there is any way to get the navigation and code highlighting with third party deps. 

That should work fine, do you have an example where it doesn't work? Is it both with Metals and Intellij?

**Reply by yadavan88** [reply]
Here is the screenshot
<img width="1230" alt="image" src="https://github.com/VirtusLab/scala-cli/assets/9131956/f85ae3be-7a5f-4286-8ac8-c692859f1747">

I have different java sample modules [here](https://github.com/yadavan88/blog-code-samples/tree/main/scala-cli-sample/scala-cli-java)

In IntelliJ, the navigation for Java core classes works. But in VS Code(with metals), no navigation and shows the errors as well.

**Reply by tgodzik** [reply]
Hmm.... Are the errors coming from Metals or the Java plugin? It does seem to work for me:

![test-java](https://github.com/VirtusLab/scala-cli/assets/3807253/44f34a42-68d9-4932-bb02-139713ac3ebe)

Though Java files will not cause the Metals server to actually start

**Reply by yadavan88** [reply]
Oh.. I am still not able to get the highlighting? 

This is the error from metals.log


```
2024.07.02 10:45:53 INFO  Attempting to connect to the build server...
2024.07.02 10:45:53 INFO  No build server found
2024.07.02 10:45:53 INFO  time: Connected to build server in 21ms
2024.07.02 10:46:14 WARN  no build tool detected in workspace '/Users/yadu/blog-code-samples/scala-cli-sample/scala-cli-java/deps'. The most common cause for this problem is that the editor was opened in the wrong working directory, for example if you use sbt then the workspace directory should contain build.sbt. 
2024.07.02 10:46:14 WARN  Build server is not auto-connectable.
2024.07.02 10:46:47 INFO  Attempting to connect to the build server...
2024.07.02 10:46:47 INFO  No build server found
2024.07.02 10:46:47 INFO  time: Connected to build server in 3ms
Jul 02, 2024 10:46:59 AM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError
SEVERE: Internal error: java.lang.Exception: No build server for target IDs file:/Users/yadu/blog-code-samples/scala-cli-sample/scala-cli-java/deps/.scala-build/?id=deps_c1106e7edc, file:/Users/yadu/blog-code-samples/scala-cli-sample/scala-cli-java/deps/.scala-build/?id=deps_c1106e7edc-test
java.util.concurrent.CompletionException: java.lang.Exception: No build server for target IDs file:/Users/yadu/blog-code-samples/scala-cli-sample/scala-cli-java/deps/.scala-build/?id=deps_c1106e7edc, file:/Users/yadu/blog-code-samples/scala-cli-sample/scala-cli-java/deps/.scala-build/?id=deps_c1106e7edc-test
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347)
	at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:708)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162)
	at scala.concurrent.java8.FuturesConvertersImpl$CF.apply(FutureConvertersImpl.scala:29)
	at scala.concurrent.java8.FuturesConvertersImpl$CF.apply(FutureConvertersImpl.scala:26)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:484)
	at scala.concurrent.ExecutionContext$parasitic$.execute(ExecutionContext.scala:222)
	at scala.concurrent.impl.Promise$Transformation.submitWithValue(Promise.scala:429)
	at scala.concurrent.impl.Promise$DefaultPromise.submitWithValue(Promise.scala:338)
	at scala.concurrent.impl.Promise$DefaultPromise.tryComplete0(Promise.scala:285)
	at scala.concurrent.impl.Promise$DefaultPromise.linkRootOf(Promise.scala:347)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:471)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.Exception: No build server for target IDs file:/Users/yadu/blog-code-samples/scala-cli-sample/scala-cli-java/deps/.scala-build/?id=deps_c1106e7edc, file:/Users/yadu/blog-code-samples/scala-cli-sample/scala-cli-java/deps/.scala-build/?id=deps_c1106e7edc-test
	at scala.meta.internal.metals.Compilations.clean$1(Compilations.scala:158)
	at scala.meta.internal.metals.Compilations.$anonfun$recompileAll$6(Compilations.scala:178)
	at scala.concurrent.Future$.$anonfun$traverse$1(Future.scala:870)
	at scala.collection.IterableOnceOps.foldLeft(IterableOnce.scala:727)
	at scala.collection.IterableOnceOps.foldLeft$(IterableOnce.scala:721)
	at scala.collection.AbstractIterator.foldLeft(Iterator.scala:1303)
	at scala.concurrent.Future$.traverse(Future.scala:870)
	at scala.meta.internal.metals.Compilations.recompileAll(Compilations.scala:177)
	at scala.meta.internal.metals.MetalsLspService.cleanCompile(MetalsLspService.scala:1870)
	at scala.meta.internal.metals.WorkspaceLspService.$anonfun$executeCommand$22(WorkspaceLspService.scala:833)
	at scala.meta.internal.metals.WorkspaceLspService.$anonfun$onCurrentFolder$5(WorkspaceLspService.scala:330)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:470)
	... 3 more


```

**Reply by yadavan88** [reply]
This is from intellij
<img width="1490" alt="image" src="https://github.com/VirtusLab/scala-cli/assets/9131956/377f9ece-6608-4db1-a841-288bdcae26e7">

I can navigate to `Exception`, but still shows errors for gson

**Reply by tgodzik** [reply]
did you try running `scala-cli setupIde .` before starting? Looks like there is some issue with which directory is opened. There should be .scala-build directory. 

The error in Metals might also be related, but it seems we do try to make something work but fail. Which directory do you open? Did you run Scala CLI before?

**Reply by yadavan88** [reply]
Yes, I did set-up ide already. 
I deleted the dir to see if there was some difference. 

I copied this file into `~/sample`. Ran `scala-cli setup-ide .` inside that directory.
Here is the ls output
<img width="349" alt="image" src="https://github.com/VirtusLab/scala-cli/assets/9131956/0b18862e-f110-48a9-85ea-40ab405d3100">

Opened it in vscode
<img width="1472" alt="image" src="https://github.com/VirtusLab/scala-cli/assets/9131956/3b97a1ef-b6f9-433d-81d7-4154d14a0fd4">

**Reply by yadavan88** [reply]
And this is from IntelliJ. BSP reload done,
<img width="1253" alt="image" src="https://github.com/VirtusLab/scala-cli/assets/9131956/2f407b93-ad9f-415c-bc86-2312ba248a1c">

**Reply by tgodzik** [reply]
Would you be able to send the snippet? Are these errors indeed from Metals not Java extension? Could you send .metals/metals.log? I would like to get to the bottom of it :thinking:

**Reply by yadavan88** [reply]
sure, how can I share the logs?

**Reply by yadavan88** [reply]
This is the metals log from the new directory. 
[metals.log](https://github.com/user-attachments/files/16066620/metals.log)

Scala-CLI version: `1.4.0`
JDK Version in the machine: `JDK 11`
Machine & OS: `MacBook Pro - M3 - MacOs Sonoma 14.5`

Here is the source code:

```
//> using jvm 17
//> using dep com.google.code.gson:gson:2.8.9

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import java.util.List;
import com.google.gson.JsonDeserializer;
import com.google.gson.GsonBuilder;

record CountryRecord(String name, String code, String capital, List<String> languages) {

}

public class JavaAppWithDep {

    public static void main(String args[]) throws Exception {

        // Custom deserializer for CountryRecord
        JsonDeserializer<CountryRecord> deserializer = (json, typeOfT, context) -> {
            var jsonObject = json.getAsJsonObject();
            String name = jsonObject.get("name").getAsString();
            String code = jsonObject.get("code").getAsString();
            String capital = jsonObject.get("capital").getAsString();
            List<String> languages = context.deserialize(jsonObject.get("languages"), new TypeToken<List<String>>() {
            }.getType());
            return new CountryRecord(name, code, capital, languages);
        };

        HttpClient httpClient = HttpClient.newHttpClient();
        URI uri = URI.create("https://raw.githubusercontent.com/yadavan88/blog-code-samples/main/countries.json");
        HttpRequest httpRequest = HttpRequest.newBuilder()
                .uri(uri)
                .build();
        HttpResponse<String> httpResponse = httpClient.send(httpRequest, HttpResponse.BodyHandlers.ofString());
        Gson gson = new GsonBuilder()
                .registerTypeAdapter(CountryRecord.class, deserializer)
                .create();
        List<CountryRecord> countries = gson.fromJson(httpResponse.body(), new TypeToken<List<CountryRecord>>() {
        }.getType());
        countries.forEach(System.out::println);
    }
}

```

**Reply by tgodzik** [reply]
Looks to be working:

![test-again](https://github.com/VirtusLab/scala-cli/assets/3807253/6dc44b3a-1d95-4738-98f3-11d56921bc62)

Maybe the issue is that somehow we are picking up JDK 11 and trying to compile with release flag?

Scala CLI actually fails to compile in your logs. `scala-cli compile .` works for you?

**Reply by tgodzik** [reply]
Btw. Metals doesn't support outline which we can see in your screenshot, that is almost 99% coming from Java extension

**Reply by tgodzik** [reply]
Although, Scala CLI does fail, which we can see in the logs :thinking:  If you disable the Java extension for the workspace does it change anything?

**Reply by tgodzik** [reply]
And also record seems to be marked with an error, which would be an error if the Java extension picked up JDK 11 from Java home

**Reply by yadavan88** [reply]
Awesome! 
Thanks for the tip. The culprit was the vscode suggested "Extension Pack for Java" extension. :|  Uninstalled it and it works great now :) 
However,  one more question, it doesn't show the "run" popup, i can see that it is showing for you. I did the setup-ide already, but still the same. 
I was wondering one more thing,  why it didn't work in IntelliJ IDEA?

<img width="1472" alt="image" src="https://github.com/VirtusLab/scala-cli/assets/9131956/a1c6de54-dbd8-4f09-aea1-8fcd8c7b011b">

**Reply by yadavan88** [reply]
Still there is warning for `record` keyword. May be it still takes from the JAVA_HOME in vs code?
Edit: I tried setting local java to 17, but still get the error for record 
It shows this tooltip in vscode:  "error: records are not supported in -source 11 bloop"

**Reply by yadavan88** [reply]
Just in case if it helps, this is the metals.log again. I cleared all the files. Reset ide initialization using setup-ide. 
Clicked on import, reconnect etc from metals.
[metals.log](https://github.com/user-attachments/files/16078628/metals.log)

**Reply by tgodzik** [reply]
> Still there is warning for record keyword. May be it still takes from the JAVA_HOME in vs code?

Looks related to https://github.com/VirtusLab/scala-cli/pull/2985 <- bloop is most likely getting JAVA_HOME from the environment at the start of the Bloop server. Instead we will use the home set by scala-cli now. Changing JAVA_HOME globally and restarting Bloop (`scala-cli bloop exit` might need the power option)

Intellij is another matter, I have no idea yet, maybe you need to import the project via BSP? There were separate import modes I think. @Gedochao was thinking to look into it

**Reply by yadavan88** [reply]
Thanks It seems the issue with Java record gone away after the laptop restart. Probably the same bloop issue.
I thought it should be working the same in intellij, let me give it a try again later tomo. I can share the logs from intellij if it helps.

**Reply by yadavan88** [reply]
Regarding the intellij, is there any specific thing i need to do to import via BSP?

**Reply by tgodzik** [reply]
Hmm.... not sure @Gedochao ?

**Reply by Gedochao** [reply]
https://scala-cli.virtuslab.org/docs/cookbooks/ide/intellij
I mean... you need to follow the standard guide for using Scala CLI with IntelliJ, as described in the docs.
And since the IDEA BSP implementation is in the Scala plugin, I suppose you need to have the Scala plugin enabled in your IntelliJ instance (even if you're just importing a pure Java project, yes).
<img width="1362" alt="image" src="https://github.com/VirtusLab/scala-cli/assets/18601388/862d0494-b82e-45fc-8cbb-754b9adbd9f8">
Seems to work as expected on my machine.
